mllib/src/main/scala/org/apache/spark/ml/classification/NaiveBayes.scala

```diff
@@ -0,0 +1,188 @@
+    override val uid: String,
+    val labels: Array[Double],
+    val pi: Array[Double],
+    val theta: Array[Array[Double]],
+    val modelType: String)
+  extends PredictionModel[Vector, NaiveBayesModel] {
+
+  /** String name for multinomial model type. */
+  private[classification] val Multinomial: String = "multinomial"
+
+  /** String name for Bernoulli model type. */
+  private[classification] val Bernoulli: String = "bernoulli"
+
+  /** Set of modelTypes that NaiveBayes supports */
+  private[classification] val supportedModelTypes = Set(Multinomial, Bernoulli)
+
+  private val piVector = new DenseVector(pi)
+  private val thetaMatrix = new DenseMatrix(labels.length, theta(0).length, theta.flatten, true)
+
+  require(supportedModelTypes.contains(modelType),
+    s"NaiveBayes was created with an unknown modelType: ${modelType}.")
+
+  /**
+   * Bernoulli scoring requires log(condprob) if 1, log(1-condprob) if 0.
+   * This precomputes log(1.0 - exp(theta)) and its sum which are used for the linear algebra
+   * application of this condition (in predict function).
+   */
+  private val (thetaMinusNegTheta, negThetaSum) = modelType match {
+    case Multinomial => (None, None)
+    case Bernoulli =>
+      val negTheta = thetaMatrix.map(value => math.log(1.0 - math.exp(value)))
+      val ones = new DenseVector(Array.fill(thetaMatrix.numCols){1.0})
+      val thetaMinusNegTheta = thetaMatrix.map { value =>
+        value - math.log(1.0 - math.exp(value))
+      }
+      (Option(thetaMinusNegTheta), Option(negTheta.multiply(ones)))
+    case _ =>
+      // This should never happen.
+      throw new UnknownError(s"Invalid modelType: ${modelType}.")
+  }
+
+  override protected def predict(features: Vector): Double = {
+    modelType match {
+      case Multinomial =>
+        val prob = thetaMatrix.multiply(features)
+        BLAS.axpy(1.0, piVector, prob)
+        labels(prob.argmax)
+      case Bernoulli =>
+        features.foreachActive{ (index, value) =>
+          if (value != 0.0 && value != 1.0) {
+            throw new SparkException(
+              s"Bernoulli naive Bayes requires 0 or 1 feature values but found ${features}")
+          }
+        }
+        val prob = thetaMinusNegTheta.get.multiply(features)
+        BLAS.axpy(1.0, piVector, prob)
+        BLAS.axpy(1.0, negThetaSum.get, prob)
+        labels(prob.argmax)
+      case _ =>
+        // This should never happen.
+        throw new UnknownError(s"Invalid modelType: ${modelType}.")
+    }
+  }
+
+  override def copy(extra: ParamMap): NaiveBayesModel = {
+    copyValues(new NaiveBayesModel(uid, labels, pi, theta, modelType), extra)
+  }
+
+  override def toString: String = {
+    s"NaiveBayesModel with ${labels.size} classes"
+  }
+
+  private[ml] def toOld: OldNaiveBayesModel = {
+    new OldNaiveBayesModel(labels, pi, theta, modelType)
+  }
+
+}
+
+private[ml] object NaiveBayesModel {
+
+  /** (private[ml]) Convert a model from the old API */
```

Comment:
This used to be necessary because of issues with generating Java docs, but you no longer need to put ```(private[ml])``` in the Scala doc.